tls/handshake.py

```diff
@@ -0,0 +1,403 @@
+        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_RCVTIMEO,
+                             struct.pack('ll', 2, self.rto * 1000))
+        if self.chunk:
+            # Send data immediately w/o coalescing.
+            self.sock.setsockopt(socket.SOL_TCP, socket.TCP_NODELAY, 1)
+        self.sock.connect((self.addr, self.port))
+
+    def send_recv(self, pkt):
+        """
+        Mainly a copy&paste from tls_do_round_trip(), but uses custom timeout to
+        be able to fully read all data from Tempesta in verbose debugging mode
+        (serial console verbose logging may be extremely slow).
+        """
+        assert self.sock, "Try to read and write on invalid socket"
+        resp = tls.TLS()
+        try:
+            if self.chunk:
+                prev_timeout = self.sock._s.gettimeout()
+                self.sock._s.settimeout(self.rto)
+                if self.sock.ctx.must_encrypt:
+                    __s = str(tls.tls_to_raw(pkt, self.sock.tls_ctx, True,
+                                             self.sock.compress_hook,
+                                             self.sock.pre_encrypt_hook,
+                                             self.sock.encrypt_hook))
+                else:
+                    __s = str(pkt)
+                n = self.chunk
+                for chunk in [__s[i:i + n] for i in xrange(0, len(__s), n)]:
+                    """
+                    This is a simple and ugly way to send many TCP segments,
+                    but it's the most applicable with other ways - some of
+                    them however gives desired effect, but in too complex way:
+                    1. SO_MAX_PACING_RATE and fair queue (TC) policing cares
+                       about total rate - it sends full TCP segments and
+                       introduces huge delays to satisfy the policy.
+                    2. TCP_MAXSEG socket option as well as reducing TCP read
+                       and write memory like
+                         sysctl -w net.ipv4.tcp_wmem='16 16 32'
+                         sysctl -w net.ipv4.tcp_rmem='16 16 32'
+                       negatively affect the whole system settings (pet socket
+                       setting also influenced by the global sysctls) and do
+                       not work with GSO & GRO - a peer still recives large
+                       frags.
+                    3. iptables --set-mss allows to change SYN packets only.
+                    4. Setting interface MTU also doesn't work properly against
+                       segmentation offloads.
+                    """
+                    self.sock._s.sendall(chunk)
+                    sleep(0.001)
+                self.sock.tls_ctx.insert(pkt, self.sock._get_pkt_origin('out'))
+                self.sock._s.settimeout(prev_timeout)
+            else:
+                self.sock.sendall(pkt, timeout=self.rto)
+            resp = self.sock.recvall(timeout=self.rto)
```

Comment:
Socket timeouts raises only when the operation (send or receive) can not be completed during the specifier time frame, so we do not introduce delays here. However, there are many timeouts around the code and some of them are really nasty, e.g. `dmeg.wait_for_msg()` - we don't know exactly when a monitored operation finishes and worse the message can be just lost, so we reach the timeout.

I also don't like the timeouts, especualy in debug mode. I created task https://github.com/tempesta-tech/tempesta-test/issues/118 . After the PR I will make another PR for performance profiling. At some point we'll face performance problem of the tests and will need to rethink timeouts as well.

Generally timeouts in tests are inevitable, at least for reliable test suite. The motivation (see [software testing and quality assurance theory and practice](https://www.wiley.com/en-us/Software+Testing+and+Quality+Assurance%3A+Theory+and+Practice-p-9780471789116) e.g. chapter 10 TEST GENERATION FROM FSM MODELS) is that the SUT is expectable misbehaving - it could hung, miss request, and so on - and the test suite must correctly handle all the misbehaving scenarios instead of hanging with the SUT.

Reply:
Here is the code of recvall:
```python
    def recvall(self, size=8192, timeout=0.5):
        resp = []
        prev_timeout = self._s.gettimeout()
        self._s.settimeout(timeout)
        while True:
            try:
                data = self._s.recv(size)
                if not data:
                    break
                resp.append(data)
            except socket.timeout:
                break
        self._s.settimeout(prev_timeout)
        records = TLS("".join(resp), ctx=self.tls_ctx, _origin=self._get_pkt_origin('in'))
        return records
```

As you can see, there `recv()` is called in a loop. There are two ways out from the loop. Either `recv()` returning empty result, which happens when other side has closed the connection. Or on read timeout. So if Tempesta responds, but doesn't close the connection, we wait for at least `timeout` seconds.

You have a relatively short timeout of 0.5 seconds, that's why it looks like there are no delays. What I was saying is that if we set timeout to something like 10 seconds, tests will slow down. But if we don't do that, tests will fail on debug builds.